warp/backenddb.go

```diff
@@ -0,0 +1,267 @@
+// (c) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package warp
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ava-labs/avalanchego/database"
+	"github.com/ava-labs/avalanchego/database/prefixdb"
+	"github.com/ava-labs/avalanchego/database/versiondb"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/hashing"
+	"github.com/ava-labs/avalanchego/utils/linkedhashmap"
+	avalancheWarp "github.com/ava-labs/avalanchego/vms/platformvm/warp"
+)
+
+var _ WarpDb = &warpDb{}
+
+var (
+	tsPrefix			= []byte("t")
+	messageIDPrefix		= []byte("m")
+)
+
+var (
+	keyExistsError = errors.New("key exists")
+)
+
+const (
+	tsSize 			= 8
+	messageIDSize 	= 32
+
+	Nanosecond 		= uint64(time.Nanosecond)
+	Microsecond 	= uint64(time.Microsecond)
+	Millisecond 	= uint64(time.Millisecond)
+	Second 			= uint64(time.Second)
+	Minute			= uint64(time.Minute)
+	Hour			= uint64(time.Hour)
+	Day				= Hour * 24
+	Week			= Day * 7
+)
+
+type prunedMap linkedhashmap.LinkedHashmap[int, []byte]
+
+func NewPrunedMap() prunedMap {
+	return linkedhashmap.New[int, []byte]()
+}
+
+type warpDbConfig struct {
+	autroprune		bool
+	maxPruneAmt		uint
+}
+
+func GetDefaultWarpDbConfig() warpDbConfig {
+	return warpDbConfig{
+		true,
+		10,
+	}
+}
+
+// WarpDb is a db that contains n messages, and automatically deletes
+// the oldest entries to make space for new ones
+type WarpDb interface {
+	AddMessage(unsignedMessage *avalancheWarp.UnsignedMessage) (int, prunedMap, error)
+	GetUnsignedMessage(messageID ids.ID) ([]byte, []byte, error)
+	PruneEntries(amt uint) (int, prunedMap, error)
+}
+
+type warpDb struct {
+	basedb			*versiondb.Database
+	msgdb			database.Database 	//maps messageID -> unsignedMessage
+	tsdb			database.Database	//maps ts -> messageID, to keep track of old messages
+	lock			sync.RWMutex
+	config			warpDbConfig			
+
+	count			uint64	//tracks total number of entries
+	timeLimit		uint64	//maximum time limit for an entry to be relevant
+}
+
+func NewWarpDb(
+	db			database.Database,
+	timeLimit	uint64,
+	config		warpDbConfig,
+) WarpDb {
+	w := warpDb{
+		count: 			0,
+		timeLimit:		timeLimit,
+	}
+	w.basedb		= versiondb.New(db)
+	w.msgdb			= prefixdb.New(messageIDPrefix, w.basedb)
+	w.tsdb 			= prefixdb.New(tsPrefix, w.basedb)
+	w.config = config
+	return &w
```

Comment:
can we clean this up so that we declare the `basedb` first and then return the whole `&warpDb...` ?